From 89e0bbf4f88bccfdcd3bd5995b749dcb8f3781f1 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Tue, 10 Aug 2010 05:42:35 -0700
Subject: [PATCH] tracing: fix oops for powerpc in irqsoff_tracer

When CONFIG_IRQSOFF_TRACER enabled, oops will be seen in lttng test
and strace.

The root cause is that only privilege mode is taken into account to
get CALLER_ADDR1 in trace_hardirqs_on by toolchain, and it caters
for IRQSOFF_TRACER.

So this fix will passby IRQSOFF_TRACER when restore to user-mode.

Signed-off-by: Liming Wang <liming.wnag@windriver.com>
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 arch/powerpc/kernel/entry_32.S |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/kernel/entry_32.S b/arch/powerpc/kernel/entry_32.S
index c6b7197..14d620a 100644
--- a/arch/powerpc/kernel/entry_32.S
+++ b/arch/powerpc/kernel/entry_32.S
@@ -928,6 +928,8 @@ native_restore:
 	 * we could encode what the exception did somewhere or test the exception
 	 * type in the pt_regs but that sounds overkill
 	 */
+	andi.	r10,r9,MSR_PR
+	bne	1f /* Do nothing if restore to user-mode */
 	andi.	r10,r9,MSR_EE
 	beq	1f
 	bl	trace_hardirqs_on
-- 
1.6.5.2

